File: Example/DigitsMigrationHelper-Example/ViewController.m
//
//  Copyright (c) 2017 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import "ViewController.h"
#import <DigitsKit/DigitsKit.h>

#import <DigitsMigrationHelper/Migrator.h>
#import <Firebase/Firebase.h>

@interface ViewController ()
@property(strong, nonatomic) FIRAuthStateDidChangeListenerHandle handleAuthStateDidChange;
@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  DGTAuthenticateButton *authButton;
  authButton = [DGTAuthenticateButton buttonWithAuthenticationCompletion:^(DGTSession *session, NSError *error) {
    if (session.userID) {
      // TODO: associate the session userID with your user model
      NSString *msg = [NSString stringWithFormat:@"Phone number: %@", session.phoneNumber];
      UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"You are logged in!"
                                                      message:msg
                                                     delegate:nil
                                            cancelButtonTitle:@"OK"
                                            otherButtonTitles:nil];
      [alert show];
    } else if (error) {
      NSLog(@"Authentication error: %@", error.localizedDescription);
    }
  }];
  authButton.center = self.view.center;
  [self.view addSubview:authButton];

  self.handleAuthStateDidChange = [[FIRAuth auth] addAuthStateDidChangeListener:^(FIRAuth *_Nonnull auth, FIRUser *_Nullable user) {
    NSLog(@"user=%@", user);
  }];
}

- (IBAction)convertDigitsUserClicked:(id)sender {
  NSLog(@"Going to convert digits user");

  NSDictionary *creds = [NSDictionary dictionaryWithContentsOfFile:[[NSBundle mainBundle] pathForResource:@"creds" ofType:@"plist"]];
  NSString *key = creds[@"consumerKey"];
  NSString *secret = creds[@"consumerSecret"];

  // [START convert_digits_user]
  FIRDigitsMigrator *migrator = [[FIRDigitsMigrator alloc] initWithDigitsAppConsumerKey:key
                                                            withDigitsAppConsumerSecret:secret];

  [migrator getLegacyAuth:^(NSString *customSignInToken, FIRDigitsSession *session){
    if (customSignInToken) {
      NSLog(@"Legacy digits session detected: token=%@, session=%@", customSignInToken, [FIRDigitsSession describe:session]);

      [[FIRAuth auth] signInWithCustomToken:customSignInToken
                                 completion:^(FIRUser * _Nullable user, NSError * _Nullable error) {
        NSLog(@"signInWithCustomToken, user=%@, error=%@", user, error);
        if (!error) {
          NSString *msg = [NSString stringWithFormat:@"Converted to firebase user with id as %@", user.uid];
          // [START_EXCLUDE silent]
          UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Conversion was successful!"
                                                          message:msg
                                                         delegate:nil
                                                cancelButtonTitle:@"OK"
                                                otherButtonTitles:nil];
          [alert show];
          // [END_EXCLUDE]

          [migrator clearLegacyAuth:^(BOOL success, NSError * _Nullable error){
            NSLog(@"clearLegacyAuth, success=%d, error=%@", success, error);
          }];
        }
      }];
    } else {
      NSLog(@"No legacy digits session detected");
      // [START_EXCLUDE]
      NSString *msg = @"Digits session not exist yet. Log into digits first to proceed.";
      UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"No digits session!"
                                                      message:msg
                                                     delegate:nil
                                            cancelButtonTitle:@"OK"
                                            otherButtonTitles:nil];
      [alert show];
      // [END_EXCLUDE]
    }
  }];
  // [END convert_digits_user]
}

- (void)viewWillDisappear:(BOOL)animated {
  [[FIRAuth auth] removeAuthStateDidChangeListener:_handleAuthStateDidChange];
}

@end

File: Example/DigitsMigrationHelper-ExampleSwift/ViewController.swift
//
//  Copyright (c) 2017 Google Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import UIKit
import DigitsMigrationHelper
import Firebase
import DigitsKit

class ViewController: UIViewController {

  var handleAuthStateDidChange: AuthStateDidChangeListenerHandle?

  override func viewDidLoad() {
    super.viewDidLoad()
    let authButton = DGTAuthenticateButton.init { (session, error) in
      if let error = error {
        print("Authentication error: \(error.localizedDescription)")
      }
      if let session = session {
        // TODO: associate the session userID with your user model
        let msg = "Phone number: \(session.phoneNumber)"
        let alert = UIAlertView.init(title: "You are logged in!", message: msg, delegate: nil, cancelButtonTitle: "OK")
        alert.show()
      }
    }
    view.addSubview(authButton!)

    handleAuthStateDidChange = Auth.auth().addStateDidChangeListener({ (auth, user) in
      print("user=\(String(describing: user))")
    })
  }

  @IBAction func convertDigitsUserClicked(sender: Any) {
    print("Going to convert digits user")

    let creds = NSDictionary.init(contentsOfFile: Bundle.main.path(forResource: "creds", ofType: "plist")!)
    let key = creds?["consumerKey"]
    let secret = creds?["consumerSecret"]

    // [START convert_digits_user]
    let migrator = DigitsMigrator.init(digitsAppConsumerKey: key as! String, withDigitsAppConsumerSecret: secret as! String)

    migrator.getLegacyAuth { (customSignInToken, session) in
      if let customSignInToken = customSignInToken {
        print("Legacy digits session detected: token=\(customSignInToken), session=\(DigitsSession.describe(session))")

        Auth.auth().signIn(withCustomToken: customSignInToken) { (user, error) in
          if let error = error {
            print("signInWithCustomToken, user=\(String(describing: user)), error=\(error)")
            return
          }

          // [START_EXCLUDE silent]
          let msg = "Converted to firebase user with id as \(user?.uid ?? "")"
          let alert = UIAlertView.init(title: "Conversion was successful!", message: msg, delegate: nil, cancelButtonTitle: "OK")
          alert.show()
          // [END_EXCLUDE]
          migrator.clearLegacyAuth { (success, error) in
            print("clearLegacyAuth, success=\(success), error=\(error?.localizedDescription ?? "")")
          }
        }
      }
    }
    // [END convert_digits_user]
  }

  override func viewDidDisappear(_ animated: Bool) {
    guard let handleAuthStateDidChange = handleAuthStateDidChange else { return }
    Auth.auth().removeStateDidChangeListener(handleAuthStateDidChange)
  }

}

